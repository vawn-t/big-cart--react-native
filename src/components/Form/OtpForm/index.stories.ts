import type { Meta, StoryObj } from '@storybook/react-native';

import OtpForm from '.';

const meta: Meta<typeof OtpForm> = {
  title: 'Components/OtpForm',
  component: OtpForm,
};

export default meta;

type Story = StoryObj<React.ComponentProps<typeof OtpForm>>;

export const Base: Story = {
  args: {},
};
