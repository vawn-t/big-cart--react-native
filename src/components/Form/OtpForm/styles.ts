import { StyleSheet } from 'react-native';
import { colors } from '@themes/index';

const styles = StyleSheet.create({
  container: { rowGap: 15 },
  form: { rowGap: 10 },
  inputWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  input: {
    backgroundColor: colors.background.primary,
    width: 55,
    height: 60,
    fontSize: 20,
    borderRadius: 5,
    textAlign: 'center',
  },
  helperContent: {
    alignItems: 'center',
  },
});

export default styles;
