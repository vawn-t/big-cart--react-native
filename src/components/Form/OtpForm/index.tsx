import { useState, useRef, useCallback, useEffect } from 'react';
import { TextInput, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import Toast from 'react-native-root-toast';

// Themes
import { colors } from '@themes/index';

// Components
import { Button, Typography } from '@components/common';

// Types
import { FontWeight, RootStackParamList, TypoVariant } from '@interfaces/index';

// Constants
import { SUCCESS, ScreenNames } from '@constants/index';

// Hooks
import useAuthVerifyCode from '@hooks/useAuth/useAuthVerifyCode';
import useAuthResendOTP from '@hooks/useAuth/useAuthResendOtp';

// Stores
import useStore from '@stores/useStore';

// Styles
import styles from './styles';

interface IProps {
  length?: number;
}

const OtpInput = ({ length = 6 }: IProps) => {
  const navigation =
    useNavigation<
      NativeStackNavigationProp<RootStackParamList, ScreenNames.VerifyNumber>
    >();

  const [otp, setOtp] = useState<string[]>(Array(length).fill(''));

  const inputRefs = useRef<TextInput[]>([]);

  // Queries
  const {
    mutateAsync: verifyCodeMutate,
    isPending: isVerifyCodePending,
    isSuccess: isVerifyCodeSuccess,
    isError: isVerifyCodeError,
    error: verifyCodeError,
  } = useAuthVerifyCode();

  const {
    mutateAsync: resendOtpMutate,
    isSuccess: isResendOtpSuccess,
    error: resendOtpError,
  } = useAuthResendOTP();

  // Stores
  const phone = useStore.use.phone();

  const [isAllFieldsFilled, setIsAllFieldsFilled] = useState(false);

  useEffect(() => {
    if (isVerifyCodeSuccess) {
      Toast.show(SUCCESS.SIGNUP.VERIFIED);

      navigation.navigate(ScreenNames.Login);
    } else if (isVerifyCodeError) {
      if (verifyCodeError.response?.data.errors?.length) {
        Toast.show(verifyCodeError.response.data.errors[0].msg);
      } else {
        Toast.show(verifyCodeError.message);
      }
    }
  }, [isVerifyCodeSuccess, isVerifyCodeError, verifyCodeError]);

  useEffect(() => {
    if (isResendOtpSuccess) {
      Toast.show(SUCCESS.RESEND_OTP);
    } else if (resendOtpError) {
      Toast.show(resendOtpError.message);
    }
  }, [isResendOtpSuccess, resendOtpError, resendOtpError]);

  useEffect(() => {
    // Auto submit form when all fields are filled
    if (isAllFieldsFilled) {
      verifyCodeMutate({ code: otp.join('') });
    }
  }, [isAllFieldsFilled]);

  const handleOtpChange = (index: number, value: string) => {
    const newOtp = [...otp];
    newOtp[index] = value;

    setOtp(newOtp);

    if (index < length - 1 && value) {
      inputRefs.current[index + 1].focus();
    }

    const allFieldsFilled = newOtp.every((value) => value !== '');
    setIsAllFieldsFilled(allFieldsFilled);
  };

  const handleOtpKeyPress = (index: number, key: string) => {
    if (key === 'Backspace' && index > 0) {
      // Moves the cursor to the previous input
      inputRefs.current[index - 1].focus();

      // Sets selection to the previous input
      inputRefs.current[index - 1].setNativeProps({
        selection: {
          start: 0,
          end: 1,
        },
      });
    }
  };

  const handleNextButton = () => {
    if (isAllFieldsFilled) {
      verifyCodeMutate({ code: otp.join('') });
    }
  };

  const handleResendCode = useCallback(() => {
    if (phone) {
      resendOtpMutate({ phone });
    }
  }, [phone]);

  return (
    <View style={styles.container}>
      <View style={styles.form}>
        <View style={styles.inputWrapper}>
          {Array(length)
            .fill('')
            .map((_, index) => (
              <TextInput
                key={index}
                ref={(ref) => (inputRefs.current[index] = ref as TextInput)}
                style={styles.input}
                maxLength={1}
                onChangeText={(value) => handleOtpChange(index, value)}
                onKeyPress={({ nativeEvent: { key } }) =>
                  handleOtpKeyPress(index, key)
                }
                value={otp[index]}
                placeholder="●"
                placeholderTextColor={colors.text.secondary}
                keyboardType="numeric"
              />
            ))}
        </View>
        <Button
          title="Next"
          onPress={handleNextButton}
          disabled={!isAllFieldsFilled || isVerifyCodePending}
        />
      </View>

      <View style={styles.helperContent}>
        <Typography
          text={`Didn't receive the code ?`}
          variant={TypoVariant.Paragraph2}
        />

        <Typography
          text="Resend a new code"
          variant={TypoVariant.Paragraph2}
          fontWeight={FontWeight.Medium}
          onPress={handleResendCode}
        />
      </View>
    </View>
  );
};

export default OtpInput;
