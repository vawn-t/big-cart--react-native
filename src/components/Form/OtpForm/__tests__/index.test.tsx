import OtpForm from '..';
import { render, fireEvent } from '@utils/index';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

const setup = () => {
  return render(<OtpForm />);
};

describe('OtpInput component', () => {
  it('renders correctly', () => {
    const { toJSON } = setup();
    expect(toJSON()).toMatchSnapshot();
  });

  xit('updates the OTP input when the user types in it', () => {
    const { getAllByPlaceholderText } = setup();

    // Find all the OTP input elements by placeholder text
    const otpInputs = getAllByPlaceholderText('●');

    // Simulate typing in each input
    otpInputs.forEach((input) => {
      fireEvent.changeText(input, '1'); // You can change the value as needed
    });

    otpInputs.forEach((_, index) => {
      expect(otpInputs[index]).toHaveProp('value', '1');
    });
  });
});
