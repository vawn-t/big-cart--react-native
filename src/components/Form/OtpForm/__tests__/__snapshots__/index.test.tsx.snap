// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OtpInput component renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "rowGap": 15,
              }
            }
          >
            <View
              style={
                {
                  "rowGap": 10,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <TextInput
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onKeyPress={[Function]}
                  placeholder="●"
                  placeholderTextColor="#000000"
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "fontSize": 20,
                      "height": 60,
                      "textAlign": "center",
                      "width": 55,
                    }
                  }
                  value=""
                />
                <TextInput
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onKeyPress={[Function]}
                  placeholder="●"
                  placeholderTextColor="#000000"
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "fontSize": 20,
                      "height": 60,
                      "textAlign": "center",
                      "width": 55,
                    }
                  }
                  value=""
                />
                <TextInput
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onKeyPress={[Function]}
                  placeholder="●"
                  placeholderTextColor="#000000"
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "fontSize": 20,
                      "height": 60,
                      "textAlign": "center",
                      "width": 55,
                    }
                  }
                  value=""
                />
                <TextInput
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onKeyPress={[Function]}
                  placeholder="●"
                  placeholderTextColor="#000000"
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "fontSize": 20,
                      "height": 60,
                      "textAlign": "center",
                      "width": 55,
                    }
                  }
                  value=""
                />
                <TextInput
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onKeyPress={[Function]}
                  placeholder="●"
                  placeholderTextColor="#000000"
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "fontSize": 20,
                      "height": 60,
                      "textAlign": "center",
                      "width": 55,
                    }
                  }
                  value=""
                />
                <TextInput
                  keyboardType="numeric"
                  maxLength={1}
                  onChangeText={[Function]}
                  onKeyPress={[Function]}
                  placeholder="●"
                  placeholderTextColor="#000000"
                  style={
                    {
                      "backgroundColor": "#FFFFFF",
                      "borderRadius": 5,
                      "fontSize": 20,
                      "height": 60,
                      "textAlign": "center",
                      "width": 55,
                    }
                  }
                  value=""
                />
              </View>
              <View
                accessibilityState={
                  {
                    "disabled": true,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "alignItems": "center",
                      "backgroundColor": "#6CC51D",
                      "borderRadius": 5,
                      "flexDirection": "row",
                    },
                    {
                      "opacity": 0.5,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    {
                      "flexGrow": 1,
                      "paddingVertical": 20,
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#000000",
                        },
                        {
                          "fontFamily": "Poppins-SemiBold",
                        },
                        {
                          "fontSize": 15,
                        },
                        {
                          "color": "#FFFFFF",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Next
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#000000",
                    },
                    {
                      "fontFamily": "Poppins-Regular",
                    },
                    {
                      "fontSize": 15,
                    },
                    undefined,
                  ]
                }
              >
                Didn't receive the code ?
              </Text>
              <Text
                onPress={[Function]}
                style={
                  [
                    {
                      "color": "#000000",
                    },
                    {
                      "fontFamily": "Poppins-Medium",
                    },
                    {
                      "fontSize": 15,
                    },
                    undefined,
                  ]
                }
              >
                Resend a new code
              </Text>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="test"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
